File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/MobzyCommands.java
package com.offz.spigot.mobzy;

import com.offz.spigot.mobzy.GUI.MobzyGUI;
import com.offz.spigot.mobzy.Mobs.Types.FlyingMob;
import com.offz.spigot.mobzy.Mobs.Types.HostileMob;
import com.offz.spigot.mobzy.Mobs.Types.PassiveMob;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.*;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

public class MobzyCommands implements org.bukkit.command.CommandExecutor, TabCompleter {
    private MobzyContext context;

    MobzyCommands(MobzyContext context) {
        this.context = context;
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!command.getName().equalsIgnoreCase("mobzy") || args.length < 1)
            return false;

        String subCommand = args[0];

        if (sender.hasPermission("mobzy.cfginfo") && subCommand.equalsIgnoreCase("cfginfo")) {
            sender.sendMessage(ChatColor.GREEN + "LOG OF CURRENTLY REGISTERED STUFF:");
            Mobzy plugin = JavaPlugin.getPlugin(Mobzy.class);
            MobzyConfig config = plugin.getMobzyConfig();
            sender.sendMessage("Mob configs: " + config.getMobCfgs());
            sender.sendMessage("Spawn configs: " + config.getSpawnCfgs());
            sender.sendMessage("Registered addons: " + config.getRegisteredAddons());
            sender.sendMessage("Registered EntityTypes: " + CustomType.getTypes());
        }

        if (sender.hasPermission("mobzy.reload") && subCommand.equalsIgnoreCase("reload")) {
            context.getMobzyConfig().reload();
            sender.sendMessage("Reloaded config files (not necessarily successfully) :p");
            return true;
        }
        List<World> worlds = context.getPlugin().getServer().getWorlds();
        boolean info = subCommand.equalsIgnoreCase("info") || subCommand.equalsIgnoreCase("i");
        if (sender.hasPermission("mobzy.remove") && (subCommand.equalsIgnoreCase("remove") || subCommand.equalsIgnoreCase("rm")) || info) {
            if (args.length < 2) {
                sender.sendMessage(ChatColor.RED + "Please specify entity mob type");
                return true;
            }
            int mobCount = 0;
            int entityCount = 0;
            for (World world : worlds)
                for (Entity entity : world.getEntities()) {
                    Set<String> tags = entity.getScoreboardTags();
                    net.minecraft.server.v1_13_R2.Entity nmsEntity = MobzyAPI.toNMS(entity);
                    if (MobzyAPI.isCustomMob(entity)
                            && ((args[1].equalsIgnoreCase("all") && !MobzyAPI.isRenamed(entity) && !entity.getScoreboardTags().contains("npc"))
                            || (args[1].equalsIgnoreCase("named") && MobzyAPI.isRenamed(entity))
                            || (args[1].equalsIgnoreCase("npc") && entity.getScoreboardTags().contains("npc"))
                            || (args[1].equalsIgnoreCase("passive") && !entity.getScoreboardTags().contains("npc") && nmsEntity instanceof PassiveMob)
                            || (args[1].equalsIgnoreCase("hostile") && nmsEntity instanceof HostileMob)
                            || (args[1].equalsIgnoreCase("flying") && nmsEntity instanceof FlyingMob)
                            || MobzyAPI.isMobOfType(entity, args[1])))
                        try {
                            Location playerLoc = Bukkit.getPlayer(sender.getName()).getLocation();
                            if (args.length < 3 || entity.getWorld().equals(playerLoc.getWorld()) && entity.getLocation().distance(playerLoc) < Integer.parseInt(args[2])) {
                                if (!info) entity.remove(); //only kill mobs if command was cmrm and not cminfo
                                entityCount++;
                                if (!tags.contains("additionalPart")) mobCount++;
                            }
                        } catch (NumberFormatException e) {
                            sender.sendMessage(ChatColor.RED + "Please input entity valid integer as the range");
                            return true;
                        }
                }
            String message = ChatColor.GREEN + "";
            if (info) message += "There are ";
            else message += "Removed ";
            message += ChatColor.BOLD + "" + mobCount + "" + ChatColor.RESET + ChatColor.GREEN;
            if (args[1].equalsIgnoreCase("all"))
                message += " custom mobs ";
            else
                message += " " + args[1] + " ";
            if (entityCount != mobCount)
                message += "(" + entityCount + " entities) ";

            if (args.length < 3) message += "in all loaded chunks";
            else message += " in entity radius of " + args[1] + " blocks";
            sender.sendMessage(message);
            return true;
        }

        if (sender.hasPermission("mobzy.spawn") && (subCommand.equalsIgnoreCase("spawn") || subCommand.equalsIgnoreCase("s"))) {
            Player p = Bukkit.getPlayer(sender.getName());
            if (args.length == 1) {
                sender.sendMessage(ChatColor.RED + "Enter a mob name");
                return true;
            }

            if (args[1] != null) {
                if (!CustomType.getTypes().containsKey(args[1])) {
                    sender.sendMessage(ChatColor.RED + "No such entity " + args[1]);
                    return true;
                }
                int numOfSpawns = 1;
                try {
                    if (args.length == 3 && args[2] != null)
                        numOfSpawns = Integer.parseInt(args[2]);

                    if (numOfSpawns > MobzyConfig.getMaxSpawnAmount())
                        numOfSpawns = MobzyConfig.getMaxSpawnAmount();

                    for (int i = 0; i < numOfSpawns; i++)
                        CustomType.spawnEntity(args[1], p.getLocation());
                } catch (NumberFormatException e) {
                    sender.sendMessage(ChatColor.RED + args[2] + " is not a valid number");
                }

            } else
                sender.sendMessage(ChatColor.RED + "Invalid mob name");
            return true;
        }

        if (sender.hasPermission("mobzy.spawn.list") && (subCommand.equalsIgnoreCase("list") || subCommand.equalsIgnoreCase("l"))) {
            sender.sendMessage(ChatColor.GREEN + CustomType.getTypes().keySet().toString());
            return true;
        }

        if (sender.hasPermission("mobzy.config") && (subCommand.equalsIgnoreCase("config"))) {
            if (args.length == 1) {
                sender.sendMessage(ChatColor.RED + "Enter a config option");
                return true;
            }
            if (args[1].equals("spawns")) {
                if (sender instanceof Player) {
                    Player player = (Player) sender;
                    new MobzyGUI(player, JavaPlugin.getPlugin(Mobzy.class)).show(player);
                }
                return true;
            }
        }

        return false;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        if (!command.getName().equalsIgnoreCase("mobzy"))
            return Collections.emptyList();

        if (args.length <= 1)
            return Stream.of("spawn", "info", "remove", "reload", "i", "rm", "s", "config")
                    .filter(a -> a.startsWith(args[0]))
                    .collect(toList());

        String subCommand = args[0];
        String cmdName = command.getName();
        if (subCommand.equalsIgnoreCase("spawn") || subCommand.equalsIgnoreCase("s"))
            if (args.length == 2) {
                return CustomType.getTypes().keySet().stream()
                        .filter(a -> a.startsWith(args[1].toLowerCase()))
                        .collect(toList());
            } else if (args.length == 3) {
                int min = 1;

                try {
                    min = Integer.parseInt(args[2]);
                } catch (NumberFormatException e) {
                }
                return IntStream.range(min, MobzyConfig.getMaxSpawnAmount() + 1)
                        .boxed()
                        .map(Object::toString)
                        .collect(toList());
            }
        if (subCommand.equalsIgnoreCase("remove") || subCommand.equalsIgnoreCase("rm") || subCommand.equalsIgnoreCase("info") || subCommand.equalsIgnoreCase("i"))
            if (args.length == 2) {
                List<String> mobs = new ArrayList<>();
                mobs.addAll(CustomType.getTypes().keySet());
                mobs.addAll(Arrays.asList("all", "npc", "mob", "named", "passive", "hostile", "flying"));
                return mobs.stream()
                        .filter(a -> a.toLowerCase().startsWith(args[1].toLowerCase()))
                        .collect(toList());
            }
        if (subCommand.equalsIgnoreCase("config"))
            return Stream.of("mobs", "spawns")
                    .filter(a -> a.toLowerCase().startsWith(args[1].toLowerCase()))
                    .collect(toList());

        return Collections.emptyList();
    }
}

File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/MobzyCommands.kt
package com.offz.spigot.mobzy

import com.mineinabyss.idofront.commands.Command.Execution
import com.mineinabyss.idofront.commands.IdofrontCommandExecutor
import com.mineinabyss.idofront.entities.toNMS
import com.mineinabyss.idofront.messaging.error
import com.mineinabyss.idofront.messaging.info
import com.mineinabyss.idofront.messaging.success
import com.offz.spigot.mobzy.gui.MobzyGUI
import com.offz.spigot.mobzy.mobs.types.FlyingMob
import com.offz.spigot.mobzy.mobs.types.HostileMob
import com.offz.spigot.mobzy.mobs.types.PassiveMob
import me.libraryaddict.disguise.utilities.DisguiseUtilities
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class MobzyCommands internal constructor(private val context: MobzyContext) : IdofrontCommandExecutor() {
    override val commands = commands(mobzy) {
        command("mobzy", "mz") {
            //the variables are stored for the commands below only
            command("configinfo", "cfginfo") {
                onExecute {
                    val config = mobzy.mobzyConfig
                    sender.info("LOG OF CURRENTLY REGISTERED STUFF:")
                    sender.info(("Mob configs: ${config.mobCfgs}\n" +
                            "Spawn configs: ${config.spawnCfgs}\n" +
                            "Registered addons: ${config.registeredAddons}\n" +
                            "Registered EntityTypes: ${mobzy.customTypes.types}"))
                }
            }
            command("reload", "rl") {
                onExecute {
                    context.mobzyConfig.reload()
                    sender.info("Reloaded config files (not necessarily successfully) :p")
                }
            }
            command("libsdisguisesstats") {
                onExecute {
                    sender.info("${DisguiseUtilities.getDisguises().count()} disguises in use")
                }
            }

            commandGroup {
                val entityType = StringArgument(1, "entity type")
                val radius = IntArgument(2, "radius", default = 0).apply { ensureChangedByPlayer() } //TODO cleaner way of doing this

                fun Execution.removeOrInfo(isInfo: Boolean) {
                    val worlds = mobzy.server.worlds
                    var mobCount = 0
                    var entityCount = 0

                    for (world in worlds) for (entity in world.entities) {
                        val tags = entity.scoreboardTags
                        val nmsEntity = entity.toNMS()
                        if (entity.isCustomMob
                                && (entityType() == "all" && !entity.isRenamed && !entity.scoreboardTags.contains("npc")
                                        || entityType() == "named" && entity.isRenamed
                                        || entityType() == "npc" && entity.scoreboardTags.contains("npc")
                                        || entityType() == "passive" && !entity.scoreboardTags.contains("npc") && nmsEntity is PassiveMob
                                        || entityType() == "hostile" && nmsEntity is HostileMob
                                        || entityType() == "flying" && nmsEntity is FlyingMob
                                        || entity.isOfType(entityType()))) {
                            val playerLoc = (sender as Player).location
                            if (radius() <= 0 || entity.world == playerLoc.world && entity.location.distance(playerLoc) < radius()) {
                                if (!isInfo) entity.remove() //only kill mobs if command was cmrm and not cminfo
                                entityCount++
                                if (!tags.contains("additionalPart")) mobCount++
                            }
                        }
                    }

                    sender.success((if (isInfo) "There are " else "Removed ") +
                            "&l$mobCount&r&a " + (if (entityType() == "all") "custom mobs " else "${entityType()} ") +
                            (if (entityCount != mobCount) "($entityCount entities) " else "") + //account for multi-entity mobs
                            (if (radius() <= 0) "in all loaded chunks." else "in a radius of ${radius()} blocks."), '&'
                    )
                }

                command("remove", "rm") {
                    onExecute { removeOrInfo(false) }
                }

                command("info", "i") {
                    onExecute { removeOrInfo(true) }
                }
            }

            command("spawn", "s") {
                val mobName = StringArgument(1, "mob name") //FIXME the function literal with receiver in onExecute passes null into the getValue method
                var numOfSpawns = IntArgument(2, "number of mobs to spawn", default = 1)

                onlyIfSenderIsPlayer()

                onExecute {
                    if (!mobzy.customTypes.types.containsKey(mobName())) { //TODO a way to add custom checks and errors via the delegation
                        sender.error("No such entity ${mobName()}")
                        return@onExecute
                    }

                    if (numOfSpawns() > MobzyConfig.maxSpawnAmount) numOfSpawns set MobzyConfig.maxSpawnAmount
                    for (i in 0 until numOfSpawns()) (sender as Player).location.spawnEntity(mobName())
                }
            }

            command("list", "l") {
                onExecute {
                    sender.success("All registered types:\n${mobzy.customTypes.types.keys}")
                }
            }
            command("config") {
                command("spawns") {
                    onlyIfSenderIsPlayer()
                    onExecute {
                        //TODO make an onExecutePlayerOnly which casts and registers the condition for us
                        MobzyGUI(sender as Player).show(sender as Player)
                    }
                }
                command("domobspawns") {
                    val enabled = BooleanArgument(1, "enabled")
                    onExecute {
                        MobzyConfig.doMobSpawns = enabled()
                        sender.success("Config option doMobSpawns has been set to ${enabled()}")
                    }
                }
            }
        }
    }

    //TODO make the API do tab completion
    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<String>): List<String> {
        if (command.name != "mobzy") return emptyList()
        if (args.size <= 1) return listOf("spawn", "info", "remove", "reload", "fullreload", "i", "rm", "s", "config", "libsdisguisesstats")
                .filter { it.startsWith(args[0]) }
        val subCommand = args[0]
        if (subCommand == "spawn" || subCommand == "s")
            if (args.size == 2) {
                return mobzy.customTypes.types.keys
                        .filter { it.startsWith(args[1].toLowerCase()) }
            } else if (args.size == 3) {
                var min = 1
                try {
                    min = args[2].toInt()
                } catch (e: NumberFormatException) {
                }
                return (min until MobzyConfig.maxSpawnAmount).asIterable()
                        .map { it.toString() }.filter { it.startsWith(min.toString()) }
            }
        if (subCommand in listOf("remove", "rm", "info", "i"))
            if (args.size == 2) {
                val mobs: MutableList<String> = ArrayList()
                mobs.addAll(mobzy.customTypes.types.keys)
                mobs.addAll(listOf("all", "npc", "mob", "named", "passive", "hostile", "flying"))
                return mobs.filter { it.toLowerCase().startsWith(args[1].toLowerCase()) }
            }
        return if (subCommand == "config") listOf("mobs", "spawns", "domobspawns")
                .filter { it.toLowerCase().startsWith(args[1].toLowerCase()) }
        else emptyList()
    }
}